src/Network/Wai/Middleware/OpenApi.hs

```diff
@@ -0,0 +1,159 @@
+module Network.Wai.Middleware.OpenApi
+  ( validate
+  , RequestErrors (..)
+  , ResponseErrors (..)
+  , SchemaNotFound (..)
+  , ValidationErrors (..)
+  , validateRequests
+  , defaultOnRequestErrors
+  , evaluateOnRequestErrors
+  , validateResponses
+  , defaultOnResponseErrors
+  , evaluateOnResponseErrors
+  ) where
+
+import Prelude
+
+import Control.Lens ((^.))
+import Control.Monad.Except
+import Control.Monad.State
+import Data.Aeson (Value, eitherDecode)
+import Data.ByteString.Builder (toLazyByteString)
+import Data.ByteString.Lazy qualified as BSL
+import Data.IORef (atomicModifyIORef, newIORef, readIORef)
+import Data.List.NonEmpty qualified as NE
+import Data.OpenApi (Definitions, OpenApi, Schema)
+import Data.OpenApi qualified as OpenApi
+import Network.Wai (Middleware, Request, Response)
+import Network.Wai qualified as Wai
+import Network.Wai.Middleware.OpenApi.PathMap qualified as PathMap
+import Network.Wai.Middleware.OpenApi.Schema
+import Network.Wai.Middleware.OpenApi.Validate
+import Network.Wai.Middleware.OpenApi.ValidationError
+
+-- | Validate using 'defaultOnRequestErrors' and 'defaultOnResponseErrors'
+validate :: OpenApi -> Middleware
+validate spec =
+  validateRequests spec defaultOnRequestErrors
+    . validateResponses spec defaultOnResponseErrors
+
+data RequestErrors
+  = RequestSchemaNotFound SchemaNotFound
+  | RequestIsNotJson BSL.ByteString String
+  | RequestInvalid Value ValidationErrors
```

Comment:
I am guessing that the `ByteString` and `Value` are the unparsed and parsed request bodies, respectively, and the `String` is a JSON parse error? Haddocks would be cool